dotnet/src/Connectors/Connectors.Amazon/Bedrock/Extensions/BedrockKernelBuilderExtensions.cs

```diff
@@ -0,0 +1,575 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+
+using Amazon;
+using Amazon.BedrockRuntime;
+using Amazon.Runtime;
+using Connectors.Amazon.Services;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.SemanticKernel;
+using Microsoft.SemanticKernel.ChatCompletion;
+using Microsoft.SemanticKernel.Connectors.Amazon.Services;
+using Microsoft.SemanticKernel.TextGeneration;
+
+namespace Connectors.Amazon.Extensions;
+
+/// <summary>
+/// Extensions for adding Bedrock services to the application.
+/// </summary>
+public static class BedrockKernelBuilderExtensions
```

Comment:
The extension methods should support the `IAmazonBedrockRuntime` being injected into the service collections. We have AWSSDK.Extensions.NETCore.Setup package that is commonly used for adding AWS service clients to the DI container and provides mechanisms for configuring the service client through the IConfiguration. Users will likely expect to be able to configure the service client there the same as they always do.

You might want to take a dependency on `AWSSDK.Extensions.NETCore.Setup` and if the user doesn't explicit give you an `IAmazonBedrockRuntime` then add it to the DI container using `AWSSDK.Extensions.NETCore.Setup`. Here is an example of where do this in another of our libraries. https://github.com/aws/aws-ssm-data-protection-provider-for-aspnet/blob/master/src/Amazon.AspNetCore.DataProtection.SSM/ExtensionMethods.cs#L58

I would also not have all the permutations of taking credentials or region or both. Have the methods take either the `IAmazonBedrockRuntime` or rely on the service client being configured via `AWSSDK.Extensions.NETCore.Setup`.

Reply:
Great, I now have the following two methods (same is done for TextGeneration). Followed the example you provided but if this looks incorrect let me know:
```
    /// <summary>
    /// Add Amazon Bedrock Chat Completion service to the kernel builder using IAmazonBedrockRuntime object.
    /// </summary>
    /// <param name="builder">The kernel builder.</param>
    /// <param name="modelId">The model for chat completion.</param>
    /// <param name="bedrockApi">The IAmazonBedrockRuntime to run inference using the respective model.</param>
    /// <param name="serviceId">The optional service ID.</param>
    /// <returns></returns>
    public static IKernelBuilder AddBedrockChatCompletionService(
        this IKernelBuilder builder,
        string modelId,
        IAmazonBedrockRuntime bedrockApi,
        string? serviceId = null)
    {
        builder.Services.AddSingleton<IChatCompletionService>(services =>
        {
            try
            {
                return new BedrockChatCompletionService(modelId, bedrockApi);
            }
            catch (Exception ex)
            {
                throw new KernelException($"An error occurred while initializing the BedrockChatCompletionService: {ex.Message}", ex);
            }
        });

        return builder;
    }

    /// <summary>
    /// Add Amazon Bedrock Chat Completion service to the kernel builder using new AmazonBedrockRuntimeClient().
    /// </summary>
    /// <param name="builder">The kernel builder.</param>
    /// <param name="modelId">The model for chat completion.</param>
    /// <param name="serviceId">The optional service ID.</param>
    /// <returns></returns>
    public static IKernelBuilder AddBedrockChatCompletionService(
        this IKernelBuilder builder,
        string modelId,
        string? serviceId = null)
    {
        // Add IAmazonBedrockRuntime service client to the DI container
        builder.Services.AddAWSService<IAmazonBedrockRuntime>();
        builder.Services.AddSingleton<IChatCompletionService>(services =>
        {
            try
            {
                var bedrockRuntime = services.GetRequiredService<IAmazonBedrockRuntime>();
                return new BedrockChatCompletionService(modelId, bedrockRuntime);
            }
            catch (Exception ex)
            {
                throw new KernelException($"An error occurred while initializing the BedrockChatCompletionService: {ex.Message}", ex);
            }
        });

        return builder;
    }
```

Reply:
I think it is better to just expect the authentication as the input from the user rather than letting them create the IAmazonBedrockRuntime object. It would be better if we expected the AWSCredentials object from the user and implement AmazonBedrockRuntimeClient(AWSCredentials, RegionEndpoint). In the future if we want to use different service endpoints i.e bedrock-runtime, bedrock-agent etc - https://docs.aws.amazon.com/bedrock/latest/APIReference/welcome.html#endpoints, we can do the same using AWSCredentials object

Reply:
I guess it makes sense that users of the semantic kernel would just want to worry about credentials, since they have to set those up. If we're making the users create the IAmazonBedrockRuntime, then aren't we kind of exposing the implementation details? We're taking a dependency on the AWS SDK now, but if it's not exposed to the users, that gives more flexibility in the future. I also could see us using other endpoints like @as14692 mentions. I guess we could support both if we have to?

Reply:
Yes, you're right I think letting user pass in their credentials makes the most sense, though I can provide the other options just so users have choices (say, in the case user wants to Mock the runtime object themselves). I previously had the Kernel Builder Extensions set up such that it allowed the user to input any of the available combination of parameters to create `new AmazonBedrockRuntimeClient(params)` including AWSCredentials, AmazonBedrockRuntimeConfig, awsAccessKeyId & awsSecretAccessKey, etc., but hadn't taken a dependency on `AWSSDK.Extensions.NETCore.Setup` to add the runtime object to the DI container. Bringing back those options and using DI would I think look like: (this is functional but not 100% sure if it's the proper implementation)
```
    public static IKernelBuilder AddBedrockChatCompletionService(
        this IKernelBuilder builder,
        string modelId,
        AWSCredentials awsCredentials)
    {
        // Create AWSOptions with the provided credentials
        var awsOptions = new AWSOptions
        {
            Credentials = awsCredentials
        };
        // Add IAmazonBedrockRuntime service client to the DI container
        builder.Services.AddAWSService<IAmazonBedrockRuntime>(awsOptions);

        builder.Services.AddSingleton<IChatCompletionService>(services =>
        {
            try
            {
                var bedrockRuntime = services.GetRequiredService<IAmazonBedrockRuntime>();
                return new BedrockChatCompletionService(modelId, bedrockRuntime);
            }
            catch (Exception ex)
            {
                throw new KernelException($"An error occurred while initializing the BedrockChatCompletionService: {ex.Message}", ex);
            }
        });

        return builder;
    }

    public static IKernelBuilder AddBedrockChatCompletionService(
        this IKernelBuilder builder,
        string modelId,
        string awsAccessKeyId,
        string awsSecretAccessKey,
        string token)
    {
        // Create AWSOptions with the provided credentials
        var awsOptions = new AWSOptions
        {
            Credentials = new SessionAWSCredentials(awsAccessKeyId, awsSecretAccessKey, token)
        };
        // Add IAmazonBedrockRuntime service client to the DI container
        builder.Services.AddAWSService<IAmazonBedrockRuntime>(awsOptions);

        builder.Services.AddSingleton<IChatCompletionService>(services =>
        {
            try
            {
                var bedrockRuntime = services.GetRequiredService<IAmazonBedrockRuntime>();
                return new BedrockChatCompletionService(modelId, bedrockRuntime);
            }
            catch (Exception ex)
            {
                throw new KernelException($"An error occurred while initializing the BedrockChatCompletionService: {ex.Message}", ex);
            }
        });

        return builder;
    }
```

Reply:
Do we need this?
```
public static IKernelBuilder AddBedrockChatCompletionService(
        this IKernelBuilder builder,
        string modelId,
        string awsAccessKeyId,
        string awsSecretAccessKey,
        string token)
```
User can create `AWSCredentials` using `awsAccessKeyId`, `awsSecretAccessKey` and `token` to call the first method. Why do we need a second implementation 🤔

Reply:
Ah, I was trying to give the user more options so they wouldn't have to create `AWSCredentials` themselves if they wanted to just pass in their keys and token. But it does makes sense to keep implementation simpler; I can just provide builder option that takes `AWSCredentials` object itself then and take that above option out.

Reply:
I would suggest not having a method that takes in credentials. Then you get into the permutation issue of you should also allow taking in region or client config. If you say it is important to provide an AWSCredentials then region and client config are just as essential. And that is the problem `AWSSDK.Extensions.NETCore.Setup` solve.

For user's that don't want to know anything about the implementation details they can call the `AddBedrockChatCompletionService` method that doesn't take any `IAmazonBedrock`. Then this library relies on AWSSDK.Extensions.NETCore.Setup for creating the instance of `IAmazonBedrock`. The override that takes in `IAmazonBedrock` is there for advanced use cases or making mocking easier.

This is the pattern we have used for the HLL the .NET SDK team has created. You configure in appsettings.json or whatever is your `IConfiguration` source the region, client config settings and aws credential profile. If those don't exist `AWSSDK.Extensions.NETCore.Setup` will use the environment look features of the SDK to determine the values.